### PR TITLE
Fix call stack size exceeded when set atom value with object that has circular reference

### LINKED ISSUE #54: Infinity loop occurs if the value to be set has a circular reference
Hi there,

I've found infinity loop problem if the value to be set has a circular reference.
Is it bug or limitation of recoil?

Minimum reproduction:
https://codesandbox.io/s/intelligent-matsumoto-vsw7m?file=/src/App.js

## Actual behavior:

```
RangeError: Maximum call stack size exceeded
```

## Expected behavior:

The value can be set without error.

### PR DESCRIPTION
**Pull request by n3tr**
## Background

fix #54

Set atom value with an object that has circular reference inside will cause an error

```
RangeError: Maximum call stack size exceeded
```

## What's changed

- In `deepFreezeValue`  Check if a property is already frozen

### CODE REVIEW

**Comment by dustinsoftware** [comment]
Cool! Seems like a pretty straight forward fix.

File: src/util/Recoil_deepFreezeValue.js [hunk 1]
```diff
@@ -64,7 +64,7 @@ function deepFreezeValue(value: mixed) {
 
   Object.freeze(value); // Make all properties read-only
   for (const prop in value) {
-    if (value.hasOwnProperty(prop)) {
+    if (value.hasOwnProperty(prop) && !Object.isFrozen(prop)) {
```
Comment (drarmstr) [inline src/util/Recoil_deepFreezeValue.js:67]: This breaks `deepFreezeValue()` because it is checking if the property string key is frozen instead of the property itself.  strings are always frozen so it essentially turns off all freezing of properties.  It causes the tests in `Recoil_core-test.js` to fail.  We can easily fix this by looking up the property itself.

This also isn't perfect, since if we encounter a non-recursive, previously frozen, object that contains unfrozen objects, they won't get frozen.  But, probably a small corner case not worth the extra cost of keeping track of visited objects to otherwise avoid infinite recursion.

File: src/recoil_values/__tests__/Recoil_atom-test.js [hunk 2]
```diff
@@ -55,3 +55,19 @@ test('atom can store null and undefined', () => {
   act(() => set(myAtom, 'VALUE'));
   expect(get(myAtom)).toBe('VALUE');
 });
+
+test('atom can store a circular reference object', () => {
```
Comment (drarmstr) [inline src/recoil_values/__tests__/Recoil_atom-test.js:59]: Need to update the test to fix Flow errors, I can fix that when fixing the diff.  Let's see if our github sync scripts are working now....

**Comment by drarmstr** [comment]
I've fixed the diff locally for our internal repository, but seems it doesn't update the GitHub PR.  Let's find out what happens when we approve and land it there if it then properly syncs with GitHub.

File: src/util/Recoil_deepFreezeValue.js (hunk 1 above)
Comment (n3tr) [reply src/util/Recoil_deepFreezeValue.js:67]: You are right! My bad, It should be `Object.isFrozen(value[prop])` instead.

File: src/recoil_values/__tests__/Recoil_atom-test.js (hunk 2 above)
Comment (n3tr) [reply src/recoil_values/__tests__/Recoil_atom-test.js:59]: Interesting, Flow doesn't show any errors on my end. I guess it should be a missing property error on Circular class, there should be something wrong with my setup.
